tensorflow/stream_executor/device_description.h

```diff
@@ -20,18 +20,72 @@ limitations under the License.
 #ifndef TENSORFLOW_STREAM_EXECUTOR_DEVICE_DESCRIPTION_H_
 #define TENSORFLOW_STREAM_EXECUTOR_DEVICE_DESCRIPTION_H_
 
+#include <iostream>
 #include <map>
 #include <memory>
-#include "tensorflow/stream_executor/platform/port.h"
 
+#include "tensorflow/core/lib/strings/numbers.h"
+#include "tensorflow/core/lib/strings/str_util.h"
+#include "tensorflow/core/platform/default/logging.h"
 #include "tensorflow/stream_executor/launch_dim.h"
+#include "tensorflow/stream_executor/lib/numbers.h"
+#include "tensorflow/stream_executor/lib/statusor.h"
+#include "tensorflow/stream_executor/lib/stringpiece.h"
+#include "tensorflow/stream_executor/platform/logging.h"
 #include "tensorflow/stream_executor/platform/port.h"
 
 namespace stream_executor {
 namespace internal {
 class DeviceDescriptionBuilder;
 }  // namespace internal
 
+// Device hardware version
+//
+// On CUDA platform it holds major and minor compute capability.
+//
+// On ROCm platform the major version is the AMDGCN ISA version, minor version
+// is zero.
+struct DeviceVersion {
+  // Initialize from version_name in the form of "3.5"
+  static port::StatusOr<DeviceVersion> Parse(port::StringPiece version_name) {
+    int major_part = 0, minor_part = 0;
+    size_t dot_pos = version_name.find('.');
+    CHECK(dot_pos != string::npos)
+        << "Illegal version name: [" << version_name << "]";
+    port::StringPiece major_str = version_name.substr(0, dot_pos);
+    CHECK(port::safe_strto32(major_str.data(), &major_part))
+        << "Illegal version name: [" << version_name << "]";
+    port::StringPiece minor_str = version_name.substr(dot_pos + 1);
+    CHECK(port::safe_strto32(minor_str.data(), &minor_part))
+        << "Illegal version name: [" << version_name << "]";
+    return DeviceVersion(major_part, minor_part);
```

Comment:
The idea was to return an error rather than CHECK (which is a fatal runtime assertion).

You'd probably want to return FailedPreconditionError, see stream_executor/lib/status.h.